rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+
+## Abstract
+
+A user stores data as a CAR file. This CAR contains a set of multihashes for individual pieces of data within the CAR. 
+
+The user creates a an external CAR index [CARx-index](https://github.com/web3-storage/RFC/pull/9/files) that contains tuples of (CID, offset, length) of any items they intend to make individually retrievable from the car.
+
+With an [inclusion claim](https://hackmd.io/@gozala/content-claims?utm_source=preview-mode&utm_medium=rec#Inclusion-Claims), the user who created the CAR asserts that CAR contains a given set of multihashes, transitively, via the CARx-index.
+
+After storing the CAR and CARx-index, w3s creates a [location claim](https://hackmd.io/@gozala/content-claims?utm_source=preview-mode&utm_medium=rec#Location-Claims) for the CAR file and the CARx-index. Each location claim maps an item's CID to a URL where the item can be retrieved.
+
+w3s creates a Content Claims Bucket. This is a container, stored on w3s network, that holds all content claims associated with the stored data. The location claims and inclusion claim are written to the claims bucket. The claims bucket is then stored.
+
+[IPNI](https://github.com/ipni/specs/blob/main/IPNI.md) will allow a client to lookup any of the multihashes, that are in the CARx-index, to get information sufficient to recover the claims bucket. The location claims from the claims bucket can then be used to get the CARx-index and CAR data.
+
+Information is provided to IPNI in the form of [IPNI Advertisements](https://github.com/ipni/specs/blob/main/IPNI.md#advertisements)
+that are signed by w3s and published to IPNI by w3s. Each  Advertisement 
+
+When a new advertisement is published, an [IPNI Announcement](https://github.com/ipni/specs/blob/main/IPNI.md#announcements) is sent to IPNI to trigger IPNI to fetch the new advertisement.
+
+## Introduction
+
+We publish ad-hoc batches of multihashes to IPNI. This proposal aims to align our usage of IPNI with [content-claims], by publishing an advert per [inclusion claim], and include the source claim in the IPNI Advertisement
+
+### Motivation
+
+- Align IPNI advertisement entries with CAR block sets, so that these CIDs can be queried to get the location of the CARx-index and CAR data.
+  - This exposes our block-to-car indexes. Anyone could use IPNI to find which CAR a block is in.
+  -   The context id bytes provide the CAR CID for any block look up. The CAR CID can then be used to find the CAR index via our content-claims API.
+  - We could delete the IPNI records by CAR CID if the CAR is deleted.
+- Make IPNI advertising an explicit UCAN capability that clients can invoke rather than a side-effect of bucket events
+  - With this we are free to write CARs anywhere. The user's agent invokes a `ipni/offer` capability to ask us to publish an IPNI advertisement for the blocks in their CAR.
+  - This empowers the user to opt-in or out as they need, and allows us to bill for the (small) cost of running that service.
```

Comment:
In the other thread we kind of converging on the idea that when user asks for the content state in their space they will get a location claim issued by service to the space did. Which is effectively our signed commitment to serve that content from that url through a ucan validity timeframe.

User can redelegate that to other user to allow them to read it. I proposed to make advertisement request on IPNI to simply be delegation of location claim to an IPNI did. Similarly authorization to serve content form gateway would be delegation to our gateway did etc… my point is we don’t need new capabilities this way, we simply allow users to choose whom to give read access

Reply:
Plz note that if UCAN authorization chain form a loop it effectively allows anyone to access excercise  it which is why publishing it to IPNI would allow anyone to perform reads

Reply:
The client can sign the advertisement as the provider. All that is needed is for the client to sign the advertisement using the private key that matches the ProviderID in the advertisement. If the advertisement includes a claim that the provider ID belongs to the client, then we can varify via UCAN that the client created the advertisement and we do not care who publishes it.

IPNI advertisements can be signed by one of two entities, the provider (sig must match ProviderID in ad) or the publisher (sig must match peer ID of host serving the advertisement).

Reply:
I think I was not very clear here let me try to spell out in more concrete terms

1. User asks to upload blob identified by multihash A and receives presigned PUT url.
2. User uploads blob to the pre-signed PUT url.
3. User checks blob status and get UCAN delegation for `assert/location` where subject `with` field is a provider DID (did:web:web3.storage), issuer is a provider DID and audience is space DID.
4. User decides to publish blob on IPNI. To do so they invoke delegated `assert/location` capability.
5. Invocation handler verifies UCAN chain and ensures that user indeed can make claim on behalf of provider `did:web:web3.storage` because the delegation chain therefor it publishes IPNI advertisements with ProviderID (because it corresponds to `with` field)

In other words provider delegates capability to the client to publish IPNI advertisement on it's behalf. As long as client does not escalate (e.g. commitment window) everything checks out during invocation so provider can simply sing and publish advertisement.

Reply:
Broader point I'm trying to make here is that service makes a commitment to the user that they will make content available, so it should be services reputation at stake and not the users of the service.

Reply:
OK, I understand that the service is making a commitment to the user about being able to serve content.

What does this mean in terms of data structures:
> To do so they invoke delegated assert/location capability.

Is this a [UCAN invocation](https://github.com/ucan-wg/invocation/blob/v0.2/README.md#ucan-invocation-specification-v020)? If so, does it contain the CID of the `assert/location` claim that w3s returned as a result of looking up blob status? Also, I do not see a `with` field in that spec.

Is it necessary to include CIDs for both CAR and CARx-index in invocation, or is CAR CID sufficient?

Reply:
@Gozala  When does the assert/location claim get created
> User checks blob status and get UCAN delegation for assert/location

Is the location claim created on-demand when the user checks status? Or, does it get created after the user uploads their blob? If created after upload, then what triggers the creation since we want to get away from bucket events?